### ISSUE TITLE
Cannot run Demo

### ISSUE THREAD

**Issue by olivierdullaert** [AhmedZgaren/Save#1]
Hi, Do you currently have support to run inference with the pre-trained model on a CPU? I am trying to do so but the CotLayer is GPU specific?

**Comment by AhmedZgaren** [comment]
Hello @olivierdullaert,

Thank you for your inquiry and for your interest in our model.

Currently, our model is optimized for GPU environments, and CPU-only execution is not an officially supported feature.

While we can't commit to an immediate timeline, we will post any updates on our progress here. In the meantime, community contributions are always welcome! If you'd like to explore adding CPU support, we'd be happy to offer guidance on the relevant parts of the codebase.

Thanks again for bringing this to our attention.

Best,

**Comment by olivierdullaert** [comment]
Hi @AhmedZgaren,

I believe I have adjusted the model to work on CPU. However, the results are not what I expected. I assume the pretrained weights are the ones of the model trained on the FSC-147 dataset. I have attached and image for which the model expects 20 instances. which is obviously too many.

could you maybe give some guidance on the generalization of the model?

![Image](https://github.com/user-attachments/assets/685bc2a9-c659-4fb8-8f3d-c476c9fa3d86)

**Comment by AhmedZgaren** [comment]
@olivierdullaert
Thanks for your interest in our model!

You've brought up a great point about the prediction accuracy on that specific image. Here are some thoughts on why the model might not have provided an accurate count in this particular case:

 - Limited Pre-training Data: The model was pre-trained on FSC-147, which is a relatively small dataset with a limited variety of objects and background contexts. This can affect its ability to generalize to novel scenarios.
 - Background Complexity: The complexity of the background in your image likely introduced noise and made it challenging for the model to accurately identify and count the target objects.
 - Out-of-Distribution Input: The image you tested appears to be significantly "out of distribution" compared to the data the model was trained on. Models often perform less robustly when presented with data that differs greatly from their training examples.
 - Ambiguity in Pattern Recognition: Our model is designed to identify and count the most frequent patterns it learned during training. In ambiguous images like the one you provided, it may struggle to find a clear, dominant pattern, leading to an inaccurate count.

**Comment by olivierdullaert** [comment]
Hi @AhmedZgaren, Thank you for your response, I have an additional question regarding the training of your model. In the paper You describe you augmented the FSC147 images to train the model. However, In you training function, the `dataloader` only uses the train and val csv files.
```
tr_d = CustomImageDataset('data/train.csv', 'data/images/')
val_d = CustomImageDataset('data/valid.csv', 'data/images/')
# train and valid dataloader
train_dataloader = DataLoader(tr_d, **params)
val_dataloader = DataLoader(val_d, **paramsv)`
```
If I want to train the model on the additional images, as you did in the paper, how would I augment the images and train the model on the augmented images? Thanks in advance!
